### PR TITLE
add new test for evm

### LINKED ISSUE #972: Fix merge account
https://github.com/AcalaNetwork/Acala/blob/b813733865b875176c7c9336c5515397c65ab90d/modules/currencies/src/lib.rs#L842-L843

Should not unreserve.
Review where it is used.
Add new tests.
Should updated the one in orml-currencies as well

### PR DESCRIPTION
**Pull request by xlc**
add provider for contracts
refactor contract removal logic
new tests

closes #972

- [x] more tests

### CODE REVIEW

File: modules/evm/src/lib.rs [hunk 1]
```diff
@@ -956,37 +960,28 @@ impl<T: Config> Pallet<T> {
...
+		ensure!(!contract_info.deployed, Error::<T>::ContractAlreadyDeployed);
+
+		let contract_account_id = T::AddressMapping::get_account_id(&contract);
+		T::Currency::unreserve(
+			&contract_account_id,
+			T::Currency::reserved_balance(&contract_account_id),
+		);
+		T::MergeAccount::merge_account(&contract_account_id, &who)?;
+
+		// should should trigger CallKillAccount and remove account
```
Comment (xlc) [inline modules/evm/src/lib.rs:979]:
```suggestion
		// shoul trigger CallKillAccount and remove account
```

File: modules/evm/src/lib.rs [hunk 2]
```diff
@@ -956,37 +960,28 @@ impl<T: Config> Pallet<T> {
...
+		ensure!(!contract_info.deployed, Error::<T>::ContractAlreadyDeployed);
+
+		let contract_account_id = T::AddressMapping::get_account_id(&contract);
+		T::Currency::unreserve(
+			&contract_account_id,
+			T::Currency::reserved_balance(&contract_account_id),
+		);
+		T::MergeAccount::merge_account(&contract_account_id, &who)?;
+
+		// shoul trigger CallKillAccount and remove account
```
Comment (zjb0807) [inline modules/evm/src/lib.rs:979]:
```suggestion
		// should trigger CallKillAccount and remove account
```

File: modules/evm/src/runner/handler.rs [hunk 3]
```diff
@@ -426,13 +426,25 @@ impl<'vicinity, 'config, 'meter, T: Config> HandlerT for Handler<'vicinity, 'con
...
 		self.storage_meter
 			.refund(size.saturating_add(T::NewContractExtraBytes::get()))
 			.map_err(|_| ExitError::Other("RefundStorageError".into()))?;
 
-		T::MergeAccount::merge_account(&source, &dest).map_err(|_| ExitError::Other("MergeAccountError".into()))
+		T::MergeAccount::merge_account(&source, &dest).map_err(|_| ExitError::Other("MergeAccountError".into()))?;
+
+		let res = frame_system::Pallet::<T>::dec_providers(&T::AddressMapping::get_account_id(&address));
+		if res == Ok(frame_system::DecRefStatus::Reaped) {
+			Ok(())
```
Comment (zjb0807) [inline modules/evm/src/runner/handler.rs:443]: should `Accounts::<T>::remove(address)`?

**Review by zjb0807 — APPROVED** [review]

File: modules/evm/src/lib.rs [hunk 4]
```diff
@@ -956,37 +960,28 @@ impl<T: Config> Pallet<T> {
...
+		let res = frame_system::Pallet::<T>::dec_providers(&contract_account_id);
+		ensure!(
+			res == Ok(frame_system::DecRefStatus::Reaped),
+			Error::<T>::CannotKillContract
+		);
```
Comment (zjb0807) [inline modules/evm/src/lib.rs:985]: should `AccountStorages::<T>::remove_prefix(contract);`?

Comment (xlc) [reply modules/evm/src/lib.rs:985]: The `OnKilledAccount` handler will call `remove_account` which will remove account storages.

**Review by ermalkaleci — APPROVED** [review]

**Comment by xlc** [comment]
@ermalkaleci @zjb0807 I did more refactor and updated how contract storage is handled. Can you review again.

**Comment by xlc** [comment]
I will get this merged to unblock #962, but please still take a review.

**Review by zjb0807 — APPROVED** [review]
LGTM

**Pull request merged**